### PR TITLE
Cataloguing System

### PR DESCRIPTION
**Pull request by nghianja**
1. Add Data Category
2. Add Keywords
3. Search by Keywords

### CODE REVIEW

File: service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import sg.ncl.common.jpa.AbstractEntity;
+import sg.ncl.service.data.domain.DataCategory;
+
+import javax.persistence.*;
+
+/**
+ * Created by dcsjnh on 9/5/2017.
+ */
+@Entity
+@Table(name = "data_categories", indexes = @Index(columnList = "name"))
```
Comment (ZHANGChunwang) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java]: For small table, probably we don't need to have index

File: service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+@Getter
+@Setter
+public class DataCategoryEntity extends AbstractEntity implements DataCategory {
+
+    @Id
+    @GeneratedValue
+    @Column(name = "id", nullable = false, unique = true, updatable = false)
+    private Long id;
+
+    @Column(name = "name", nullable = false, unique = true, updatable = false)
```
Comment (ZHANGChunwang) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java]: Why not allow category name to be updated?

Comment (nghianja) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java]: Currently has no interface to allow the updating of category. Will allow update when we work on the admin portion for category.

File: service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java (hunk 1 above)
Comment (nghianja) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataCategoryEntity.java]: This one is written by Te Ye in the specs so I just copied. Will remove in the next iteration of change.

File: service-data/src/main/java/sg/ncl/service/data/data/jpa/DataEntity.java [hunk 3]
```diff
@@ -64,10 +51,17 @@
...
+    @JoinColumn(name = "category_id")
+    private DataCategoryEntity category;
+
     @OneToMany(cascade = CascadeType.ALL, mappedBy = "dataEntity", orphanRemoval = true)
     private List<DataResourceEntity> resources = new ArrayList<>();
 
-
+    @ElementCollection
+    @CollectionTable(name = "data_keywords", joinColumns = @JoinColumn(name = "data_id", nullable = false, updatable = false), indexes = {@Index(columnList = "data_id"), @Index(columnList = "keyword")}, uniqueConstraints = @UniqueConstraint(columnNames = {"data_id", "keyword"}))
+    @Column(name = "keyword", nullable = false, updatable = false)
```
Comment (ZHANGChunwang) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataEntity.java]: Keywords should also be updatable (maybe not from the web interface, but from direct SQL ops)

Comment (nghianja) [inline service-data/src/main/java/sg/ncl/service/data/data/jpa/DataEntity.java]: Okay.

File: src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql [hunk 4]
```diff
@@ -0,0 +1,39 @@
+-- -----------------------------------------------------
+-- Table `prod`.`data_keywords`
+-- Keywords that describes a dataset.
+-- -----------------------------------------------------
+CREATE TABLE IF NOT EXISTS `prod`.`data_keywords` (
+  `data_id`    BIGINT(20)    NOT NULL,
+  `keyword`    VARCHAR(255)  NOT NULL,
+  PRIMARY KEY (`data_id`, `keyword`),
+  INDEX `IDXa565e755709166653ad81ac92b39cc15` (`data_id` ASC),
+  INDEX `IDXd7df5b64df1181ef1d62d646a13aa860` (`keyword` ASC),
```
Comment (ZHANGChunwang) [inline src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql]: Primary key is always indexed, so there is no need for these two indexes...

Comment (nghianja) [inline src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql]: I didn't know that. I will remove it.

File: src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+-- Table `prod`.`data_categories`
+-- Categories that a dataset can belongs to.
+-- -----------------------------------------------------
+CREATE TABLE IF NOT EXISTS `prod`.`data_categories` (
+  `id`       	         BIGINT(20)	  NOT NULL AUTO_INCREMENT,
+  `created_date`       DATETIME     NOT NULL,
+  `last_modified_date` DATETIME     NOT NULL,
+  `version`            BIGINT(20)   NOT NULL,
+  `name`               VARCHAR(255)	NOT NULL,
+  `description`        LONGTEXT	    NOT NULL,
```
Comment (ZHANGChunwang) [inline src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql]: In DataCategoryEntity, description can be NULL. To make them consistent.

Comment (nghianja) [inline src/main/resources/db/migration/V1_12__create_tables_for_catalogue.sql]: Okay.

**Pull request merged**